### PR TITLE
[#1428] fix(server): fallback invalid when local storage can't write

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?

allow specifying negative fallback threshold to avoid event being discarded

### Why are the changes needed?

Fix: #1428 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by zuston** [comment]
PTAL @leixm @jerqi . I think this is a critical bug in 0.8 branch.

**Comment by jerqi** [comment]
I can't get the point of this pr.

**Comment by zuston** [comment]
> I can't get the point of this pr.

Please see the test case

**Comment by zuston** [comment]
cc @xianjingfeng If you have time, could you help review this ?

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 1]
```diff
@@ -352,9 +352,7 @@ public class ShuffleServerConf extends RssBaseConf {
   public static final ConfigOption<Long> FALLBACK_MAX_FAIL_TIMES =
       ConfigOptions.key("rss.server.hybrid.storage.fallback.max.fail.times")
           .longType()
-          .checkValue(
-              ConfigUtils.NON_NEGATIVE_LONG_VALIDATOR, " fallback times must be non-negative")
-          .defaultValue(0L)
+          .defaultValue(-1L)
```
Comment (xianjingfeng) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:355]: 1. What if the users set it to 0.
2. If `!storageManager.canWrite(flushEvent)`, we should ignore retry times. So should we add a new argument named `force` to `org.apache.uniffle.server.storage.AbstractStorageManagerFallbackStrategy#tryFallback` for ignoring retry times. WDYT? @zuston  @jerqi 

https://github.com/apache/incubator-uniffle/blob/318050144b62c0256c151a278bb04ac73c6e544b/server/src/main/java/org/apache/uniffle/server/storage/hybrid/FallbackBasedStorageManagerSelector.java#L49-L51

Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:355]: SGTM.

Comment (jerqi) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:355]: Ok for me.

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 2]
```diff
@@ -352,8 +352,6 @@ public class ShuffleServerConf extends RssBaseConf {
   public static final ConfigOption<Long> FALLBACK_MAX_FAIL_TIMES =
       ConfigOptions.key("rss.server.hybrid.storage.fallback.max.fail.times")
           .longType()
-          .checkValue(
```
Comment (zuston) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:355]: The negative check is unnecessary

**Comment by zuston** [comment]
PTAL again @xianjingfeng

**Review by xianjingfeng — APPROVED** [review]
LGTM

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java (hunk 2 above)
Comment (jerqi) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:355]: Why do we remove this check?

Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:355]: if -1，it means we could fallback directly once failed at the first time.

Comment (jerqi) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:355]: What about the behaviour of `0`?

Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:355]: For the first time fallback, it will reject for 0

**Comment by zuston** [comment]
Merged. Thanks @xianjingfeng @jerqi

**Pull request merged**